File: app/src/main/java/edu/gatech/donatrix/controllers/ItemSearchActivity.java
package edu.gatech.donatrix.controllers;

import android.content.Intent;
import android.support.annotation.Nullable;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.View;
import android.widget.AdapterView;
import android.widget.ArrayAdapter;
import android.widget.EditText;
import android.widget.Spinner;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.Collection;
import java.util.HashMap;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;

import edu.gatech.donatrix.R;
import edu.gatech.donatrix.dao.LocationDao;
import edu.gatech.donatrix.data.RESTCaller;
import edu.gatech.donatrix.model.ItemCategory;
import edu.gatech.donatrix.model.Location;

/**
 * An Activity to search for items.
 */
public class ItemSearchActivity extends AppCompatActivity
        implements AdapterView.OnItemSelectedListener {

    private EditText nameText;
    private Spinner resultSpinner;

    @Nullable
    private Location location;
    private String category;

    private List<Location> locations;

    private String item;
    private List<Map<String, Object>> items;


    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_item_search);

        Spinner locationSpinner = findViewById(R.id.itemSearchLocationSpinner);
        Spinner categorySpinner = findViewById(R.id.itemSearchCategorySpinner);
        locationSpinner.setOnItemSelectedListener(this);
        categorySpinner.setOnItemSelectedListener(this);
        nameText = findViewById(R.id.editText);

        Log.d("Donatrix", "Hello");
        resultSpinner = findViewById(R.id.itemSearchResultSpinner);
        resultSpinner.setOnItemSelectedListener(this);

        List<Location>  locList = LocationDao.getLocations(this);
        locations = locList;
        List<String> locs = new ArrayList<>();
        for (Location loc: locList) {
            locs.add(loc.getName());
        }
        locs.add(0, "");
        ArrayAdapter<Location> locationArrayAdapter = new ArrayAdapter(
                this, android.R.layout.simple_spinner_item, locs.toArray());
        locationArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        locationSpinner.setAdapter(locationArrayAdapter);

        Log.d("Donatrix", "Hello2");
        ItemCategory[] itemCategoryArray = ItemCategory.values();
        List<String> categories = new ArrayList<>();
        for (ItemCategory i : itemCategoryArray) {
            categories.add(i.getCategory());
        }
        categories.add(0, "");
        ArrayAdapter<String> itemCategoryArrayAdapter = new ArrayAdapter(
                this, android.R.layout.simple_spinner_item, categories.toArray());
        Log.d("Donatrix", "Hello3");
        itemCategoryArrayAdapter.setDropDownViewResource(
                android.R.layout.simple_spinner_dropdown_item);
        Log.d("Donatrix", "Hello5");
        categorySpinner.setAdapter(itemCategoryArrayAdapter);
        Log.d("Donatrix", "Hello4");

        String[] result = new String[1];
        result[0] = "";
        ArrayAdapter<String> stringArrayAdapter = new ArrayAdapter(
                this, android.R.layout.simple_spinner_item, result);
        stringArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item);
        resultSpinner.setAdapter(stringArrayAdapter);
        Log.d("Donatrix", "Hello6");
    }

    @Override
    public void onItemSelected(AdapterView<?> parent, View view, int position, long id) {
        try {
            Spinner spinner = (Spinner) parent;
            if (spinner.getId() == R.id.itemSearchLocationSpinner) {
                boolean changed = false;
                for (Location loc: locations) {
                    if (loc.getName().equals(spinner.getItemAtPosition(position))) {
                        location = loc;
                        changed = true;
                    }
                }
                if (!changed) {
                    location = null;
                }
            } else if (spinner.getId() == R.id.itemSearchCategorySpinner) {
                category = (String) spinner.getItemAtPosition(position);
            } else if (spinner.getId() == R.id.itemSearchResultSpinner) {
                item = (String) spinner.getItemAtPosition(position);
            }
            Log.d("Donatrix", "Please Print");
        } catch (Exception e ) {
            Log.d("Donatrix", e.getMessage());
        }

    }

    @Override
    public void onNothingSelected(AdapterView<?> parent) {
        category = "";
        location = null;
    }

    /**
     * What to do when cancelled.
     * @param view XML
     */
    public void onCancelButtonPressed(View view) {
        finish();
    }

    /**
     * What happens when search is pressed.
     * @param view XML
     */
    public void onSearchButtonPressed(View view) {
        Map<String, Object> body = new HashMap<>();
        if (location != null) {
            body.put("loc_id", location.getKey());
        }
        if (!category.isEmpty()) {
            body.put("cat", category);
        }
        if (!("" + nameText.getText()).isEmpty()) {
            body.put("sDesc", "" + nameText.getText());
        }
        Log.d("Donatrix", body.toString());
        Map<String, Object> response = RESTCaller.post(
                "https://donatrix-api.herokuapp.com/filtered/getItems", body);
        boolean success = (boolean) response.get("success");
        if (success) {
            List<Map<String, Object>> res = (List<Map<String, Object>>) response.get("items");
            items = res;
            Collection<String> result = new ArrayList<>();
            for (Map m : res) {
                Log.d("Donatrix", m.toString());
                result.add((String) m.get("s_description"));
            }
            ArrayAdapter<String> stringArrayAdapter = new ArrayAdapter(this,
                    android.R.layout.simple_spinner_item, result.toArray());
            stringArrayAdapter.setDropDownViewResource(
                    android.R.layout.simple_spinner_dropdown_item);
            resultSpinner.setAdapter(stringArrayAdapter);
        }
    }

    /**
     * How to get the details of an item.
     * @param view XML
     */
    public void onDetailsButtonPressed(View view) {
        String itemName = (String) resultSpinner.getItemAtPosition(0);
        Map<String, Object> item = new LinkedHashMap<>();
        for (Map m: items) {
            if (m.get("s_description").equals(this.item)) {
                item = m;
            }
        }
        Intent intent = new Intent(ItemSearchActivity.this, ItemDetailActivity.class);
        intent.putExtra("item", (Serializable) item);
        startActivity(intent);
    }
}

File: app/src/main/java/edu/gatech/donatrix/controllers/ItemSearchActivity.kt
package edu.gatech.donatrix.controllers

import android.content.Intent
import android.support.v7.app.AppCompatActivity
import android.os.Bundle
import android.util.Log
import android.view.View
import android.widget.AdapterView
import android.widget.ArrayAdapter
import android.widget.EditText
import android.widget.Spinner

import java.io.Serializable
import java.util.ArrayList
import java.util.HashMap
import java.util.LinkedHashMap

import edu.gatech.donatrix.R
import edu.gatech.donatrix.dao.LocationDao
import edu.gatech.donatrix.data.RESTCaller
import edu.gatech.donatrix.model.ItemCategory
import edu.gatech.donatrix.model.Location

class ItemSearchActivity : AppCompatActivity(), AdapterView.OnItemSelectedListener {

    private var nameText: EditText? = null
    private var resultSpinner: Spinner? = null

    private var location: Location? = null
    private var category: String? = null

    private var locations: List<Location>? = null

    private var item: String? = null
    private var items: List<Map<String, Any>>? = null


    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_item_search)

        val locationSpinner = findViewById<Spinner>(R.id.itemSearchLocationSpinner)
        val categorySpinner = findViewById<Spinner>(R.id.itemSearchCategorySpinner)
        locationSpinner.onItemSelectedListener = this
        categorySpinner.onItemSelectedListener = this
        nameText = findViewById(R.id.editText)

        Log.d("Donatrix", "Hello")
        resultSpinner = findViewById(R.id.itemSearchResultSpinner)
        resultSpinner!!.onItemSelectedListener = this

        val locList = LocationDao.getLocations(this)
        locations = locList
        val locs = ArrayList<String?>()
        for (loc in locList) {
            locs.add(loc.name)
        }
        locs.add(0, "")
        val locationArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, locs.toTypedArray())
        locationArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        locationSpinner.adapter = locationArrayAdapter

        Log.d("Donatrix", "Hello2")
        val itemCategoryArray = ItemCategory.values()
        val categories = ArrayList<String>()
        for (i in itemCategoryArray) {
            categories.add(i.category)
        }
        categories.add(0, "")
        val itemCategoryArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, categories.toTypedArray())
        Log.d("Donatrix", "Hello3")
        itemCategoryArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        Log.d("Donatrix", "Hello5")
        categorySpinner.adapter = itemCategoryArrayAdapter
        Log.d("Donatrix", "Hello4")

        val result = arrayOfNulls<String>(1)
        result[0] = ""
        val stringArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, result)
        stringArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
        resultSpinner!!.adapter = stringArrayAdapter
        Log.d("Donatrix", "Hello6")
    }

    override fun onItemSelected(parent: AdapterView<*>, view: View, position: Int, id: Long) {
        try {
            val spinner = parent as Spinner
            if (spinner.id == R.id.itemSearchLocationSpinner) {
                var changed = false
                for (loc in locations!!) {
                    if (loc.name == spinner.getItemAtPosition(position)) {
                        location = loc
                        changed = true
                    }
                }
                if (!changed) {
                    location = null
                }
            } else if (spinner.id == R.id.itemSearchCategorySpinner) {
                category = spinner.getItemAtPosition(position) as String
            } else if (spinner.id == R.id.itemSearchResultSpinner) {
                item = spinner.getItemAtPosition(position) as String
            }
            Log.d("Donatrix", "Please Print")
        } catch (e: Exception) {
            Log.d("Donatrix", e.message)
        }

    }

    override fun onNothingSelected(parent: AdapterView<*>) {
        category = ""
        location = null
    }

    fun onCancelButtonPressed(view: View) {
        finish()
    }

    fun onSearchButtonPressed(view: View) {
        val body = HashMap<String, Any>()
        if (location != null) {
            body["loc_id"] = location!!.key
        }
        if (!category!!.isEmpty()) {
            body["cat"] = category!!
        }
        if (!("" + nameText!!.text).isEmpty()) {
            body["sDesc"] = "" + nameText!!.text
        }
        Log.d("Donatrix", body.toString())
        val response = RESTCaller.post("https://donatrix-api.herokuapp.com/filtered/getItems", body)
        val success = response["success"] as Boolean
        if (success) {
            val res = response["items"] as List<Map<String, Any>>?
            items = res
            val result = ArrayList<String>()
            for (m in res!!) {
                Log.d("Donatrix", m.toString())
                result.add(m["s_description"] as String)
            }
            val stringArrayAdapter = ArrayAdapter(this, android.R.layout.simple_spinner_item, result.toTypedArray())
            stringArrayAdapter.setDropDownViewResource(android.R.layout.simple_spinner_dropdown_item)
            resultSpinner!!.adapter = stringArrayAdapter
        }
    }

    fun onDetailsButtonPressed(view: View) {
        val itemName = resultSpinner!!.getItemAtPosition(0) as String
        var item: Map<String, Any> = LinkedHashMap()
        for (m in items!!) {
            if (m["s_description"] == this.item) {
                item = m
            }
        }
        val intent = Intent(this@ItemSearchActivity, ItemDetailActivity::class.java)
        intent.putExtra("item", item as Serializable)
        startActivity(intent)
    }
}
